notification-service/src/websocket.ts

```diff
@@ -0,0 +1,50 @@
+import { ethers } from 'ethers';
+import { INFURA_PROJECT_ID, ETHEREUM_NETWORK } from './variables';
+import { SUBSCRIPTIONS } from './constants/SUBSCRIPTIONS';
+import { WEBSOCKET_PREFIX } from './constants/PREFIXES';
+import { EventData, EventSubscription } from './types';
+import { getAbiFromContractAddress } from './etherscan';
+
+export function setupWebSocket(): ethers.providers.WebSocketProvider {
+    if (!INFURA_PROJECT_ID) {
+        throw new Error(`${WEBSOCKET_PREFIX} please set a valid INFURA_PROJECT_ID, in the environment files`);
+    }
+    const wsProvider = new ethers.providers.WebSocketProvider(
+        `wss://${ETHEREUM_NETWORK}.infura.io/ws/v3/${INFURA_PROJECT_ID}`,
+        ETHEREUM_NETWORK
+    );
+
+    console.info(`${WEBSOCKET_PREFIX} websocket connection opened with network "${ETHEREUM_NETWORK}"`);
+    return wsProvider;
+}
+
+export function listenForEvents(
+    wsProvider: ethers.providers.WebSocketProvider,
+    triggerEvent: (eventData: EventData) => void
+) {
+    const contracts: Record<string, EventSubscription[]> = {};
+
+    SUBSCRIPTIONS.forEach(eventSubscription => {
+        if (!contracts[eventSubscription.address]) {
+            contracts[eventSubscription.address] = [eventSubscription];
+        } else {
+            contracts[eventSubscription.address].push(eventSubscription);
+        }
+    });
+
+    Object.keys(contracts).forEach(async contractAddress => {
+        const contractAbi = await getAbiFromContractAddress(contractAddress);
```

Comment:
This function may fail because 
- network error
- no contract has no abi
- etc

Leaving the whole subscription service stale, right?

Reply:
I changed it to display the following message if the retrieving fails. I also display the Error of the function, so the user can see if it was because of one of your listed reasons.

![Screenshot 2022-05-06 at 00 07 47](https://user-images.githubusercontent.com/30908158/167033421-33165be6-55c5-4de0-befa-b4f647fd2a49.png)